Color suggestion fails with a particular shade of green

Hi!

Noticed this while using a11y-color in a side-project. Here's a repro:

I added the following to `test/color.js`:

```js
describe("with a white foreground and a light green background", () => {
  it("returns a dark green", () => {
    let fg = parseColor("rgb(255, 255, 255)");
    let bg = parseColor("rgb(2, 184, 117)");

    let suggestions = suggestColors(bg, fg, {
      AA: 3.0,
      AAA: 4.5,
    });

    expect(suggestions.AA.contrast).to.be.at.least(2.95);
    expect(suggestions.AAA.contrast).to.be.at.least(4.45);
  });
});
```

`npm test` then fails with the following:

```
  1) suggestColors with a white foreground and a light green background returns a dark green:
     AssertionError: expected '2.91' to be at least 2.95
      at Context.<anonymous> (test/color.js:91:48)
```

It seems `suggestColors` is sending me back a ratio of 2.91 and 3.91 when I ask for 3.0 and 4.5 respectively.

This also seems to be a problem in accessibility developer tools: http://jsbin.com/voluwoqezo/edit?html,js,console

--- a/gulpfile.js
+++ b/gulpfile.js
@@ -48,5 +48,5 @@ gulp.task('provide-color-src', function () {
     .pipe(gulp.dest('lib'));
 });
 
-gulp.task('default', ['jshint', 'jscs']);
+gulp.task('default', ['jshint']);
 gulp.task('compile', ['jshint', 'jscs', 'docs', 'provide-color-src']);

--- a/package.json
+++ b/package.json
@@ -17,7 +17,7 @@
   "devDependencies": {
     "babel": "^5.4.7",
     "chai": "^3.2.0",
-    "gulp": "^3.9.0",
+    "gulp": "^3.9.1",
     "gulp-jscs": "^2.0.0",
     "gulp-jsdoc": "^0.1.4",
     "gulp-jshint": "^1.11.2",

--- a/src/color.js
+++ b/src/color.js
@@ -355,9 +355,8 @@ function luminanceFromContrastRatio (luminance, contrast, higher) {
  * @private
  */
 function translateColor (ycc, luma) {
-  let endpoint = (luma > ycc.luma) ? WHITE_YCC : BLACK_YCC;
-  let cubeFaces = (endpoint == WHITE_YCC) ? YCC_CUBE_FACES_WHITE
-    : YCC_CUBE_FACES_BLACK;
+  let lighter = luma > ycc.luma;
+  let endpoint = lighter ? WHITE_YCC : BLACK_YCC;
 
   let a = new YCbCr([0, ycc.Cb, ycc.Cr]);
   let b = new YCbCr([1, ycc.Cb, ycc.Cr]);
@@ -366,34 +365,40 @@ function translateColor (ycc, luma) {
 
   let intersection = null;
 
-  for (let i = 0; i < cubeFaces.length; i++) {
-    let cubeFace = cubeFaces[i];
-    intersection = findIntersection(line, cubeFace);
+  let cubeFaces = (lighter ? YCC_CUBE_FACES_WHITE : YCC_CUBE_FACES_BLACK);
+  for (let cubeFace of cubeFaces) {
+    let candidateIntersection = findIntersection(line, cubeFace);
+    if (candidateIntersection.z < 0 || candidateIntersection.z > 1)
+      continue;
 
-    if (intersection.z >= 0 && intersection.z <= 1) {
-      break;
+    if (!intersection) {
+      intersection = candidateIntersection;
+      continue;
     }
+
+    // May intersect more than one plane, since planes continue beyond edges of cube.
+    // Find the closest intersection to original luma, as this will be the cube edge.
+    if (Math.abs(candidateIntersection.z - ycc.luma) < Math.abs(intersection.luma - ycc.luma))
+	intersection = candidateIntersection;
   }
 
-  if (!intersection) {
+  if (!intersection)
     throw "Couldn't find intersection with YCbCr color cube for Cb=" + ycc.Cb + ", Cr=" + ycc.Cr + ".";
-  }
 
-  if (intersection.x != ycc.x || intersection.y != ycc.y) {
+  if (intersection.x != ycc.x || intersection.y != ycc.y)
     throw "Intersection has wrong Cb/Cr values.";
-  }
 
-  if (Math.abs(endpoint.luma - intersection.luma) < Math.abs(endpoint.luma - luma)) {
-    let translatedColor = [luma, ycc.Cb, ycc.Cr];
+  // If luma is closer to endpoint than intersection.luma is, point is outside cube.
+  if (Math.abs(endpoint.luma - intersection.luma) > Math.abs(endpoint.luma - luma)) {
+    let dLuma = luma - intersection.luma;
+    let scale = dLuma / (endpoint.luma - intersection.luma);
+    let translatedColor = [luma,
+			   intersection.Cb - (intersection.Cb * scale),
+			   intersection.Cr - (intersection.Cr * scale)];
     return fromYCbCrArray(translatedColor);
   }
 
-  let dLuma = luma - intersection.luma;
-  let scale = dLuma / (endpoint.luma - intersection.luma);
-  let translatedColor = [luma,
-      intersection.Cb - (intersection.Cb * scale),
-      intersection.Cr - (intersection.Cr * scale)];
-
+  let translatedColor = [luma, ycc.Cb, ycc.Cr];
   return fromYCbCrArray(translatedColor);
 }
 

--- a/test/color.js
+++ b/test/color.js
@@ -105,6 +105,21 @@ describe("calculateContrastRatio", () => {
       });
     });
   });
+
+  describe("with a white background and a light green foreground", () => {
+    it("returns a dark green", () => {
+      let bg = parseColor("rgb(255, 255, 255)");
+      let fg = parseColor("rgb(2, 184, 117)");
+
+      let suggestions = suggestColors(bg, fg, {
+          AA: 3.0,
+	  AAA: 4.5
+      });
+
+      expect(suggestions.AA.contrast).to.be.at.least(2.95);
+      expect(suggestions.AAA.contrast).to.be.at.least(4.45);
+    });
+  });
 });
 
 describe("Color", () => {

